File: src/TIL_0909/kakao_04_trie.py
class Trie:
    def __init__(self):
        self.childList = [None]*26
        self.cnt = 0

    def insert(self, word):
        current = self
        for letter in word:
            current.cnt += 1
            if current.childList[ord(letter)-ord('a')]:
                current = current.childList[ord(letter)-ord('a')]
            else:
                current.childList[ord(letter)-ord('a')] = Trie()
                current = current.childList[ord(letter)-ord('a')]

    def search(self, query):
        current = self

        for i in range(0, len(query)):
            if query[i]=='?':
                return current.cnt

            if current.childList[ord(query[i])-ord('a')]:
                current = current.childList[ord(query[i])-ord('a')]
            else:
                return 0

        return current.cnt


def solution(words, queries):
    answer = []
    words = set(words)
    queryDict = {}

    tries = [None for _ in range(0, 100001)] 
    rtries = [None for _ in range(0, 100001)]

    for word in words:
        if not tries[len(word)]:
            tries[len(word)] = Trie()
        tries[len(word)].insert(word)
        if not rtries[len(word)]:
            rtries[len(word)] = Trie()
        rtries[len(word)].insert(word[::-1])

    for query in queries:
        if query not in queryDict:
            cnt = 0
            if query[0]=='?' and rtries[len(query)]:
                cnt = rtries[len(query)].search(query[::-1])
            elif tries[len(query)]:
                cnt = tries[len(query)].search(query)
        
            queryDict[query] = cnt

        answer.append(queryDict[query])

    return answer



words = ["frodo", "front", "frost", "frozen", "kakao", "frame" ]
queries = ["fro??", "????o", "fr???", "fro???", "pro?"]

print(solution(words, queries))

File: src/TIL_0909/kakao_04_trie.java
package TIL_0909;

public class kakao_04_trie {
	
	public static void main(String[] args) {
		String[] words = { "a","b","frodo", "front", "frost", "frozen", "kakao", "frame" };
		String[] queries = { "?","fro??", "????o", "fr???", "fro???", "pro??", "??ont", "????o" };
		
		System.out.println(solution(words, queries));
	}
    
	public static int[] solution(String[] words, String[] queries) {
        int[] answer = new int[queries.length];
        
        Trie[] tries = new Trie[10001];
        Trie[] rtries = new Trie[10001];
        
        for (String word : words) {
        	int size = word.length();
        	try {
        		tries[size].insert(word.toCharArray());
        	}
        	catch(Exception e){
        		tries[size] = new Trie();
        		tries[size].insert(word.toCharArray());
        	}
        	
        	word = (new StringBuffer(word).reverse().toString());
        	try {
        		rtries[size].insert(word.toCharArray());
        	}
        	catch(Exception e){
        		rtries[size] = new Trie();
        		rtries[size].insert(word.toCharArray());
        	}
        }

        for(int i=0; i<queries.length; i++) {
    		String query = queries[i];
    		if(query.indexOf("?")==0) {
    			try {
    				query = (new StringBuffer(query).reverse().toString());
    				answer[i] = rtries[query.length()].search(query.toCharArray());
    			}
    			catch(Exception e) {
    				continue;
    			}
    		}
    		else {
    			try {
    				answer[i] = tries[query.length()].search(query.toCharArray());
    			}
    			catch(Exception e) {
    				continue;
    			}
    		}
    	}        
        
        return answer;
    }
	
}

class Trie {
	
	int count;
	Trie[] childList;
	
	Trie() {
		childList = new Trie[26];
		count = 0;
	}
	
	void insert(char[] word) {
		Trie current = this;
		for (char no : word) {
			current.count++;
			if(current.childList[no - 'a'] != null){
				current = current.childList[no - 'a'];
			}
			else {
				current.childList[no - 'a'] = new Trie();
				current = current.childList[no - 'a'];
			}
		}
	}
	
	int search(char[] query) {
		Trie current = this;
		for (char no : query) {
			if (no == '?') {
				return current.count;
			}
			if (current.childList[no - 'a'] != null) {
				current = current.childList[no - 'a'];
			}
			else {
				return 0;
			}
		}
		return current.count;
		
	}
	
}
